File: lib/tier-config.ts
/** Subscription tier for PedaStudio teachers. */
export type TierId = "basic" | "prime" | "max";

export const TIER_IDS: TierId[] = ["basic", "prime", "max"];

export const TIER_LABELS: Record<TierId, string> = {
  basic: "Basic",
  prime: "Prime",
  max: "Max",
};

/** App ships Grades 1–5 English only (Balbharati). */
export const GRADES_APP = [1, 2, 3, 4, 5] as const;
const GRADES_FULL = [1, 2, 3, 4, 5];

/** Weekly AI usage caps (resets every Monday UTC). */
export const BASIC_PLANS_PER_WEEK = 2;
export const PRIME_PLANS_PER_WEEK = 6;
export const MAX_PLANS_PER_WEEK = 6;
export const PRIME_OCR_SCANS_PER_WEEK = 6;
export const MAX_SCANS_PER_WEEK = 2;
export const MAX_OCR_SCANS_PER_WEEK = 12;

/** One-time Max trial for new teachers (days). */
export const MAX_TRIAL_DAYS = 7;

export interface TierLimits {
  plansPerWeek: number | null;
  worksheetsPerWeek: number | null;
  scansPerWeek: number | null;
  maxClasses: number | null;
  maxStudentsPerClass: number;
  ocrScansPerWeek: number | null;
}

export interface TierFeatures {
  unlimitedPlans: boolean;
  planModesAlways: boolean;
  planModesAfterUnitTest: boolean;
  baselineAssessment: boolean;
  endlineAssessment: boolean;
  unitTests: boolean;
  manualAssessmentEntry: boolean;
  fullAssessmentReports: boolean;
  shortActionPlan: boolean;
  fullActionPlan: boolean;
  baselinePlanBand: boolean;
  unitTlmKit: boolean;
  tlmGapListUnit: boolean;
  yearTlmListView: boolean;
  yearTlmListFull: boolean;
  yearTlmPdfShare: boolean;
  bulkPaperScan: boolean;
  aiAutoMark: boolean;
  reportPdfExport: boolean;
  worksheets: boolean;
  textbookScan: boolean;
  perStudentLongitudinal: boolean;
  abilityGroups: boolean;
  clusterExport: boolean;
  hindiUrduUi: boolean;
  gradesAvailable: number[];
}

export const TIER_LIMITS: Record<TierId, TierLimits> = {
  basic: {
    plansPerWeek: BASIC_PLANS_PER_WEEK,
    worksheetsPerWeek: 0,
    scansPerWeek: 0,
    maxClasses: 1,
    maxStudentsPerClass: 45,
    ocrScansPerWeek: 0,
  },
  prime: {
    plansPerWeek: PRIME_PLANS_PER_WEEK,
    worksheetsPerWeek: null,
    scansPerWeek: 0,
    maxClasses: 1,
    maxStudentsPerClass: 45,
    ocrScansPerWeek: PRIME_OCR_SCANS_PER_WEEK,
  },
  max: {
    plansPerWeek: MAX_PLANS_PER_WEEK,
    worksheetsPerWeek: null,
    scansPerWeek: MAX_SCANS_PER_WEEK,
    maxClasses: 2,
    maxStudentsPerClass: 60,
    ocrScansPerWeek: MAX_OCR_SCANS_PER_WEEK,
  },
};

export const TIER_FEATURES: Record<TierId, TierFeatures> = {
  basic: {
    unlimitedPlans: false,
    planModesAlways: true,
    planModesAfterUnitTest: true,
    baselineAssessment: true,
    endlineAssessment: true,
    unitTests: true,
    manualAssessmentEntry: true,
    fullAssessmentReports: false,
    shortActionPlan: true,
    fullActionPlan: false,
    baselinePlanBand: true,
    unitTlmKit: true,
    tlmGapListUnit: true,
    yearTlmListView: true,
    yearTlmListFull: false,
    yearTlmPdfShare: false,
    bulkPaperScan: false,
    aiAutoMark: false,
    reportPdfExport: false,
    worksheets: false,
    textbookScan: false,
    perStudentLongitudinal: false,
    abilityGroups: false,
    clusterExport: false,
    hindiUrduUi: true,
    gradesAvailable: [...GRADES_FULL],
  },
  prime: {
    unlimitedPlans: false,
    planModesAlways: true,
    planModesAfterUnitTest: true,
    baselineAssessment: true,
    endlineAssessment: true,
    unitTests: true,
    manualAssessmentEntry: true,
    fullAssessmentReports: true,
    shortActionPlan: true,
    fullActionPlan: true,
    baselinePlanBand: true,
    unitTlmKit: true,
    tlmGapListUnit: true,
    yearTlmListView: true,
    yearTlmListFull: true,
    yearTlmPdfShare: true,
    bulkPaperScan: true,
    aiAutoMark: true,
    reportPdfExport: false,
    worksheets: true,
    textbookScan: false,
    perStudentLongitudinal: true,
    abilityGroups: false,
    clusterExport: false,
    hindiUrduUi: true,
    gradesAvailable: [...GRADES_FULL],
  },
  max: {
    unlimitedPlans: false,
    planModesAlways: true,
    planModesAfterUnitTest: true,
    baselineAssessment: true,
    endlineAssessment: true,
    unitTests: true,
    manualAssessmentEntry: true,
    fullAssessmentReports: true,
    shortActionPlan: true,
    fullActionPlan: true,
    baselinePlanBand: true,
    unitTlmKit: true,
    tlmGapListUnit: true,
    yearTlmListView: true,
    yearTlmListFull: true,
    yearTlmPdfShare: true,
    bulkPaperScan: true,
    aiAutoMark: true,
    reportPdfExport: true,
    worksheets: true,
    textbookScan: true,
    perStudentLongitudinal: true,
    abilityGroups: true,
    clusterExport: true,
    hindiUrduUi: true,
    gradesAvailable: [...GRADES_FULL],
  },
};

export function normalizeTierId(value: unknown): TierId {
  if (value === "prime" || value === "max" || value === "basic") return value;
  return "basic";
}

/** When set (e.g. PILOT_TIER=prime), all teachers get this tier from the API. Leave unset in production. */
export function getPilotTierOverride(): TierId | null {
  const raw = process.env.PILOT_TIER?.trim().toLowerCase();
  if (raw === "basic" || raw === "prime" || raw === "max") return raw;
  return null;
}

/** ISO date (YYYY-MM-DD) of the current week’s Monday, UTC. */
export function currentUsageWeek(): string {
  const now = new Date();
  const day = now.getUTCDay();
  const daysFromMonday = day === 0 ? 6 : day - 1;
  const monday = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate() - daysFromMonday));
  return monday.toISOString().slice(0, 10);
}

export type UsageCounterKey = "plans" | "worksheets" | "scans" | "ocrScans";

export interface UsageSnapshot {
  week: string;
  plans: number;
  worksheets: number;
  scans: number;
  ocrScans: number;
}

export function emptyUsage(week = currentUsageWeek()): UsageSnapshot {
  return { week, plans: 0, worksheets: 0, scans: 0, ocrScans: 0 };
}

File: android/app/src/main/java/com/tippingpoint/pedastudio/data/TierConfig.kt
package com.tippingpoint.pedastudio.data

/** Mirrors lib/tier-config.ts for client-side UI gating. */
object TierConfig {
    enum class TierId(val key: String, val label: String) {
        BASIC("basic", "Basic"),
        PRIME("prime", "Prime"),
        MAX("max", "Max"),
    }

    const val BASIC_PLANS_PER_WEEK = 2
    const val PRIME_PLANS_PER_WEEK = 6
    const val MAX_PLANS_PER_WEEK = 6
    const val PRIME_OCR_SCANS_PER_WEEK = 6
    const val MAX_SCANS_PER_WEEK = 2
    const val MAX_OCR_SCANS_PER_WEEK = 12

    data class Limits(
        val plansPerWeek: Int?,
        val worksheetsPerWeek: Int?,
        val scansPerWeek: Int?,
        val maxClasses: Int?,
        val maxStudentsPerClass: Int,
        val ocrScansPerWeek: Int?,
    )

    data class Features(
        val unlimitedPlans: Boolean,
        val planModesAlways: Boolean,
        val planModesAfterUnitTest: Boolean,
        val baselineAssessment: Boolean,
        val endlineAssessment: Boolean,
        val unitTests: Boolean,
        val manualAssessmentEntry: Boolean,
        val fullAssessmentReports: Boolean,
        val shortActionPlan: Boolean,
        val fullActionPlan: Boolean,
        val baselinePlanBand: Boolean,
        val unitTlmKit: Boolean,
        val tlmGapListUnit: Boolean,
        val yearTlmListView: Boolean,
        val yearTlmListFull: Boolean,
        val yearTlmPdfShare: Boolean,
        val bulkPaperScan: Boolean,
        val aiAutoMark: Boolean,
        val reportPdfExport: Boolean,
        val worksheets: Boolean,
        val textbookScan: Boolean,
        val perStudentLongitudinal: Boolean,
        val abilityGroups: Boolean,
        val clusterExport: Boolean,
        val hindiUrduUi: Boolean,
        val gradesAvailable: List<Int>,
    )

    data class Usage(
        val week: String,
        val plans: Int,
        val worksheets: Int,
        val scans: Int,
        val ocrScans: Int,
    )

    fun parseTier(raw: String?): TierId =
        when (raw?.lowercase()) {
            "prime" -> TierId.PRIME
            "max" -> TierId.MAX
            else -> TierId.BASIC
        }

    fun defaultAccount(): TeacherAccount = TeacherAccount(
        tier = TierId.BASIC,
        limits = limitsFor(TierId.BASIC),
        features = featuresFor(TierId.BASIC),
        usage = Usage(week = "", plans = 0, worksheets = 0, scans = 0, ocrScans = 0),
        plansRemaining = BASIC_PLANS_PER_WEEK,
        subscription = SubscriptionInfo("none", null, null, null),
        paymentsEnabled = false,
    )

    fun tierRank(tier: TierId): Int = when (tier) {
        TierId.BASIC -> 0
        TierId.PRIME -> 1
        TierId.MAX -> 2
    }

    fun isAtLeast(current: TierId, required: TierId): Boolean =
        tierRank(current) >= tierRank(required)

    fun limitsFor(tier: TierId): Limits = when (tier) {
        TierId.BASIC -> Limits(BASIC_PLANS_PER_WEEK, 0, 0, 1, 45, 0)
        TierId.PRIME -> Limits(PRIME_PLANS_PER_WEEK, null, 0, 1, 45, PRIME_OCR_SCANS_PER_WEEK)
        TierId.MAX -> Limits(MAX_PLANS_PER_WEEK, null, MAX_SCANS_PER_WEEK, 2, 60, MAX_OCR_SCANS_PER_WEEK)
    }

    fun featuresFor(tier: TierId): Features {
        val fullGrades = listOf(1, 2, 3, 4, 5)
        return when (tier) {
            TierId.BASIC -> Features(
                unlimitedPlans = false,
                planModesAlways = true,
                planModesAfterUnitTest = true,
                baselineAssessment = true,
                endlineAssessment = true,
                unitTests = true,
                manualAssessmentEntry = true,
                fullAssessmentReports = false,
                shortActionPlan = true,
                fullActionPlan = false,
                baselinePlanBand = true,
                unitTlmKit = true,
                tlmGapListUnit = true,
                yearTlmListView = true,
                yearTlmListFull = false,
                yearTlmPdfShare = false,
                bulkPaperScan = false,
                aiAutoMark = false,
                reportPdfExport = false,
                worksheets = false,
                textbookScan = false,
                perStudentLongitudinal = false,
                abilityGroups = false,
                clusterExport = false,
                hindiUrduUi = true,
                gradesAvailable = fullGrades,
            )
            TierId.PRIME -> Features(
                unlimitedPlans = false,
                planModesAlways = true,
                planModesAfterUnitTest = true,
                baselineAssessment = true,
                endlineAssessment = true,
                unitTests = true,
                manualAssessmentEntry = true,
                fullAssessmentReports = true,
                shortActionPlan = true,
                fullActionPlan = true,
                baselinePlanBand = true,
                unitTlmKit = true,
                tlmGapListUnit = true,
                yearTlmListView = true,
                yearTlmListFull = true,
                yearTlmPdfShare = true,
                bulkPaperScan = true,
                aiAutoMark = true,
                reportPdfExport = false,
                worksheets = true,
                textbookScan = false,
                perStudentLongitudinal = true,
                abilityGroups = false,
                clusterExport = false,
                hindiUrduUi = true,
                gradesAvailable = fullGrades,
            )
            TierId.MAX -> Features(
                unlimitedPlans = false,
                planModesAlways = true,
                planModesAfterUnitTest = true,
                baselineAssessment = true,
                endlineAssessment = true,
                unitTests = true,
                manualAssessmentEntry = true,
                fullAssessmentReports = true,
                shortActionPlan = true,
                fullActionPlan = true,
                baselinePlanBand = true,
                unitTlmKit = true,
                tlmGapListUnit = true,
                yearTlmListView = true,
                yearTlmListFull = true,
                yearTlmPdfShare = true,
                bulkPaperScan = true,
                aiAutoMark = true,
                reportPdfExport = true,
                worksheets = true,
                textbookScan = true,
                perStudentLongitudinal = true,
                abilityGroups = true,
                clusterExport = true,
                hindiUrduUi = true,
                gradesAvailable = fullGrades,
            )
        }
    }

    fun canGeneratePlan(account: TeacherAccount): Boolean =
        account.plansRemaining == null || account.plansRemaining > 0

    fun gradeAllowed(account: TeacherAccount, grade: Int): Boolean =
        account.features.gradesAvailable.contains(grade)

    fun upgradeLabel(tier: TierId): String = when (tier) {
        TierId.BASIC -> "Prime"
        TierId.PRIME -> "Max"
        TierId.MAX -> "Max"
    }
}
